flink-clients/src/test/java/org/apache/flink/client/program/PerJobMiniClusterTest.java

```diff
@@ -0,0 +1,234 @@
+
+/**
+ * Tests for {@link PerJobMiniCluster}.
+ */
+public class PerJobMiniClusterTest {
+
+	@ClassRule
+	public static TemporaryFolder temporaryFolder = new TemporaryFolder();
+
+	private MiniCluster miniCluster;
+
+	@BeforeClass
+	public static void beforeClass() throws Exception {
+		temporaryFolder.create();
+	}
+
+	@Test
+	public void testJobExecution() throws Exception {
+		PerJobMiniCluster perJobMiniCluster = initializeMiniCluster();
+
+		JobClient jobClient = perJobMiniCluster.submitJob(getNoopJobGraph()).get();
+
+		JobExecutionResult jobExecutionResult = jobClient.getJobExecutionResult(getClass().getClassLoader()).get();
+		assertThat(jobExecutionResult, is(notNullValue()));
+
+		Map<String, Object> actual = jobClient.getAccumulators(getClass().getClassLoader()).get();
+		assertThat(actual, is(notNullValue()));
+
+		assertThatMiniClusterIsShutdown();
+	}
+
+	@Test
+	public void testJobClient() throws Exception {
+		PerJobMiniCluster perJobMiniCluster = initializeMiniCluster();
+
+		JobGraph cancellableJobGraph = getCancellableJobGraph();
+		JobClient jobClient = perJobMiniCluster
+			.submitJob(cancellableJobGraph)
+			.get();
+
+		assertThat(jobClient.getJobID(), is(cancellableJobGraph.getJobID()));
+		assertThat(jobClient.getJobStatus().get(), is(JobStatus.RUNNING));
+
+		jobClient.cancel().get();
+
+		try {
+			jobClient.getJobExecutionResult(getClass().getClassLoader()).get();
+		} catch (ExecutionException e) {
+			assertThat(e.getMessage(), containsString("Failed to convert JobResult to JobExecutionResult."));
+		}
+
+		assertThatMiniClusterIsShutdown();
+	}
+
+	@Test
+	public void testJobClientSavepoint() throws Exception {
+		JobGraph jobGraph = new JobGraph();
+		String savepointPath = temporaryFolder.getRoot().getAbsolutePath();
+
+		PerJobMiniCluster perJobMiniCluster = new PerJobMiniCluster(new Configuration(), config -> {
+			// Use a mock in this test case to test calling of the savepoint methods
+			miniCluster = Mockito.mock(MiniCluster.class);
+
+			CompletableFuture<JobSubmissionResult> jobSubmissionFuture = new CompletableFuture<>();
+			jobSubmissionFuture.complete(new JobSubmissionResult(jobGraph.getJobID()));
+			Mockito.when(miniCluster.submitJob(Mockito.any())).thenReturn(jobSubmissionFuture);
+
+			CompletableFuture<JobResult> jobResultFuture = new CompletableFuture<>();
+			jobResultFuture.complete(
+				new JobResult.Builder()
+					.jobId(jobGraph.getJobID())
+					.netRuntime(42L)
+					.build());
+			Mockito.when(miniCluster.requestJobResult(Mockito.any())).thenReturn(jobResultFuture);
+
+			return miniCluster;
+		});
+
+		JobClient jobClient = perJobMiniCluster.submitJob(jobGraph).get();
+
+		jobClient.triggerSavepoint(savepointPath);
+		Mockito.verify(miniCluster).triggerSavepoint(jobGraph.getJobID(), savepointPath, false);
+
+		jobClient.stopWithSavepoint(true, savepointPath);
+		Mockito.verify(miniCluster).stopWithSavepoint(jobGraph.getJobID(), savepointPath, true);
+	}
```

Comment:
I'm not so sure about the value of this test case. We setup quite some Mockito machinery to test that `PerJobMiniClusterJobClient` forwards `triggerSavepoint` to `MiniCluster`. I think this does not justify the complexity of this test. Since we don't test `stopWithSavepoint`, I would be fine with removing this test case given that its test scope is quite limited anyway.

Maybe it would be easier to test this aspect if we introduced a new interface for representing the interaction between the `MiniCluster` and the `PerJobMiniClusterJobClient`.

Reply:
I've removed the Mockito machinery (love that term). I've replaced it with a simple test that checks whether the savepoint is triggered.

Reply:
Nice :-)